contracts/src/token/erc721/extensions/consecutive.rs

```diff
@@ -0,0 +1,1054 @@
+        )?)
+    }
+
+    fn transfer_from(
+        &mut self,
+        from: Address,
+        to: Address,
+        token_id: U256,
+    ) -> Result<(), Error> {
+        if to.is_zero() {
+            return Err(erc721::Error::InvalidReceiver(
+                ERC721InvalidReceiver { receiver: Address::ZERO },
+            )
+            .into());
+        }
+
+        // Setting an "auth" argument enables the `_is_authorized` check which
+        // verifies that the token exists (`from != 0`). Therefore, it is
+        // not needed to verify that the return value is not 0 here.
+        let previous_owner = self._update(to, token_id, msg::sender())?;
+        if previous_owner != from {
+            return Err(erc721::Error::IncorrectOwner(ERC721IncorrectOwner {
+                sender: from,
+                token_id,
+                owner: previous_owner,
+            })
+            .into());
+        }
+        Ok(())
+    }
+
+    fn approve(&mut self, to: Address, token_id: U256) -> Result<(), Error> {
+        self._approve(to, token_id, msg::sender(), true)
+    }
+
+    fn set_approval_for_all(
+        &mut self,
+        operator: Address,
+        approved: bool,
+    ) -> Result<(), Error> {
+        Ok(self.erc721.set_approval_for_all(operator, approved)?)
+    }
+
+    fn get_approved(&self, token_id: U256) -> Result<Address, Error> {
+        self._require_owned(token_id)?;
+        Ok(self.erc721._get_approved_inner(token_id))
+    }
+
+    fn is_approved_for_all(&self, owner: Address, operator: Address) -> bool {
+        self.erc721.is_approved_for_all(owner, operator)
+    }
+}
+
+// erc721 related implementation:
+impl Erc721Consecutive {
+    /// Returns the owner of the `token_id`. Does NOT revert if the token
+    /// doesn't exist.
+    ///
+    /// IMPORTANT: Any overrides to this function that add ownership of tokens
+    /// not tracked by the core [`Erc721`] logic MUST be matched with the use
+    /// of [`Self::_increase_balance`] to keep balances consistent with
+    /// ownership. The invariant to preserve is that for any address `a` the
+    /// value returned by `balance_of(a)` must be equal to the number of
+    /// tokens such that `owner_of_inner(token_id)` is `a`.
+    #[must_use]
+    fn __owner_of_inner(&self, token_id: U256) -> Address {
+        self.erc721._owners.get(token_id)
+    }
+
+    /// Transfers `token_id` from its current owner to `to`, or alternatively
+    /// mints (or burns) if the current owner (or `to`) is the `Address::ZERO`.
+    /// Returns the owner of the `token_id` before the update.
+    ///
+    /// The `auth` argument is optional. If the value passed is non-zero, then
+    /// this function will check that `auth` is either the owner of the
+    /// token, or approved to operate on the token (by the owner).
+    ///
+    /// NOTE: If overriding this function in a way that tracks balances, see
+    /// also [`Self::_increase_balance`].
+    fn __update(
```

Comment:
Why do we need `__`? Is `_update` not enough? Tbh I am confused with `__update`, `_update` etc.

Reply:
As I mentioned in the pr, part of the code is duplicated from erc721. `__update` function name highlights function copied from erc721. `_update` overrides that function. Not the best naming though.. Which approach you think could look better @bidzyyys @alexfertel ?

Reply:
I gave this another look and I think I don't understand very well why we *need* to duplicate it, since we can just call `self.erc721._update()`, right?

Reply:
I think no, since every time we call an `__update()` function `_owner_of_inner()` is called inside of it.
`_owner_of_inner()` is a bit tweaked function, that also uses `_sequential_burn` and `_sequential_ownership` storage fields. That is how I've tried to emulate solidity inheritance

Reply:
How do you think guys it can be done better?

Reply:
Yeah, I see it now. We need to use the new `_owner_of_inner`, so we can't call `self.erc721._update` because it uses the old `_owner_of_inner`.

I see `__update` only gets called once in line 408, inside `_update`. We have two options:
- Rename `__update` to something else, like for example: `_update_consecutive`.
- Inline `__update`.

wdyt?

Reply:
Yeah! As I remember `__update` was unlined before. Then I thought it would be convenient for maintenance to split unique code of consecutive and erc721 duplicated one. And extracted it as an `__update` function.
I think the public function should have the same name `_update` as erc721 base implementation has. To not surprise users.
`__update` is private right now and we are free to use any name. How do u feel about  `_update_base` or `_update_super` ?

Reply:
Okay, let's go with `_update_base`. Let's also remove the `NOTE` comment since we can't override it, and let's document that this is a copy of `_update`, but we have to duplicate it because

> We need to use the new _owner_of_inner, so we can't call self.erc721._update because it uses the old _owner_of_inner.